lua/packer/clean.lua

```diff
@@ -9,44 +9,75 @@ local async = a.sync
 
 local config
 
+local PLUGIN_OPTIONAL_LIST = 1
+local PLUGIN_START_LIST = 2
+
 local function is_dirty(plugin, typ)
-  return plugin.disable or (plugin.opt and typ == 2) or (not plugin.opt and typ == 1)
+  return plugin.disable or (plugin.opt and typ == PLUGIN_START_LIST) or (not plugin.opt and typ == PLUGIN_OPTIONAL_LIST)
 end
 
 -- Find and remove any plugins not currently configured for use
 local clean_plugins = function(_, plugins, results)
   return async(function()
+    local dirty_plugins = {}
     results = results or {}
     results.removals = results.removals or {}
+
     local opt_plugins, start_plugins = plugin_utils.list_installed_plugins()
-    local dirty_plugins = {}
-    local aliases = {}
-    for _, plugin in pairs(plugins) do
-      if plugin.as and not plugin.disable then aliases[plugin.as] = true end
+
+    local missing_plugins = plugin_utils.find_missing_plugins(plugins, opt_plugins, start_plugins)
+    -- turn the list into a hashset-like structure
+    for idx, plugin_name in ipairs(missing_plugins) do
+      missing_plugins[plugin_name] = true
+      missing_plugins[idx] = nil
     end
 
-    for typ, plugin_list in ipairs({opt_plugins, start_plugins}) do
-      for plugin_path, _ in pairs(plugin_list) do
-        local plugin_name = vim.fn.fnamemodify(plugin_path, ":t")
-        local plugin_data = plugins[plugin_name]
-        if (plugin_data == nil and not aliases[plugin_name])
-          or (plugin_data and is_dirty(plugin_data, typ)) then
-          dirty_plugins[plugin_path] = plugin_name
-        end
+	 -- test for dirty / 'missing' plugins
+    for _, plugin_config in pairs(plugins) do
+      local path = plugin_config.install_path
+
+      local plugin_source = nil
+      if opt_plugins[path] then
+        plugin_source = PLUGIN_OPTIONAL_LIST
+        opt_plugins[path] = nil
+      elseif start_plugins[path] then
+        plugin_source = PLUGIN_START_LIST
+        start_plugins[path] = nil
+      end
+
+		-- We don't want to report paths which don't exist for removal; that will confuse people
+      local plugin_missing = missing_plugins[plugin_config.short_name] and vim.loop.fs_stat(path)
+
+      if plugin_missing or is_dirty(plugin_config, plugin_source) then
+        dirty_plugins[#dirty_plugins+1] = path
```

Comment:
Nitpick: Here and at line 59, is there a reason this isn't `table.insert(dirty_plugins, path)`?

Reply:
I usually try to use the insertion syntax where possible, because it's just one less resolution for Lua to go through (see [here](https://stackoverflow.com/questions/38345582/why-is-tt-1-e-faster-than-table-insertt-e)). I did notice that `table.insert` is preferred by a lot of the Neovim community, though, and I don't mind changing this— it was just my instinct.

Reply:
Oh, cool. I didn't know that - intuitively, I thought that the insertion syntax would trigger metamethods, whereas `table.insert` would not, so I actually thought `table.insert` would be faster. Good to learn, thanks! I'm fine with using the insertion syntax (or not if you don't feel like reverting the change to `table.insert`).

Reply:
I'll leave it to match the style in the rest of the project, though we can create another performance PR if we wanted to use that syntax for appending.